erpnext/stock/doctype/pick_list/pick_list.py

```diff
@@ -314,64 +344,88 @@ def get_available_item_locations_for_other_item(item_code, from_warehouses, requ
-		else:
-			sales_order_item = None
+				dn_item = map_child_doc(source_doc, delivery_note, table_mapper)
 
-		source_doc, table_mapper = [sales_order_item, item_table_mapper] if sales_order_item \
-			else [location, item_table_mapper_without_so]
+				if dn_item:
+					dn_item.warehouse = location.warehouse
+					dn_item.qty = flt(location.picked_qty) / (flt(location.conversion_factor) or 1)
+					dn_item.batch_no = location.batch_no
+					dn_item.serial_no = location.serial_no
 
-		dn_item = map_child_doc(source_doc, delivery_note, table_mapper)
+					update_delivery_note_item(source_doc, dn_item, delivery_note)
 
-		if dn_item:
-			dn_item.warehouse = location.warehouse
-			dn_item.qty = flt(location.picked_qty) / (flt(location.conversion_factor) or 1)
-			dn_item.batch_no = location.batch_no
-			dn_item.serial_no = location.serial_no
+		set_delivery_note_missing_values(delivery_note)
 
-			update_delivery_note_item(source_doc, dn_item, delivery_note)
+		delivery_note.pick_list = pick_list.name
+		delivery_note.customer = frappe.get_value("Sales Order",sales_order,"customer")
 
-	set_delivery_note_missing_values(delivery_note)
+	pick_list = frappe.get_doc('Pick List', source_name)
+	validate_item_locations(pick_list)
+	sales_dict = dict()
+	sales_orders = []
+	delivery_note = None
+	for d in pick_list.locations:
+		if d.sales_order:
+			sales_orders.append([frappe.db.get_value("Sales Order",d.sales_order,'customer'),d.sales_order])
+
+	from itertools import groupby
+	from operator import itemgetter
+
+	# Group sales orders by customer
+	for key,keydata in groupby(sales_orders,key=itemgetter(0)):
+		sales_dict[key] = set([d[1] for d in keydata])
+
+	for customer in sales_dict:
+		for so in sales_dict[customer]:
+			delivery_note = None
+			delivery_note = create_delivery_note_from_sales_order(so,
+				delivery_note, skip_item_mapping=True)
+
+			item_table_mapper = {
+				'doctype': 'Delivery Note Item',
+				'field_map': {
+					'rate': 'rate',
+					'name': 'so_detail',
+					'parent': 'against_sales_order',
+				},
+				'condition': lambda doc: abs(doc.delivered_qty) < abs(doc.qty) and doc.delivered_by_supplier!=1
+			}
+			break;
+		if delivery_note:
+			# map all items of all sales orders of that customer
+			for so in sales_dict[customer]:		
+				map_pl_locations(pick_list.locations,so)
+			delivery_note.insert(ignore_mandatory = True)
+
+	# Create a DN for items without sales orders as well
+	if not delivery_note:
+		delivery_note = frappe.new_doc("Delivery Note")
 
-	delivery_note.pick_list = pick_list.name
-	delivery_note.customer = pick_list.customer if pick_list.customer else None
+		item_table_mapper_without_so = {
+			'doctype': 'Delivery Note Item',
+			'field_map': {
+				'rate': 'rate',
+				'name': 'name',
+				'parent': '',
+			}
+		}
+		map_pl_locations(pick_list.locations,None)
+		delivery_note.insert(ignore_mandatory = True)
 
+	frappe.msgprint('Delivery Note(s) created for the Pick List!')
 	return delivery_note
```

Comment:
create_delivery_note looks like a big function so difficult to read , could you split it into smaller functions

Reply:
Done